File: examples/node/establishments.ts
import {
  ParceleMaisClient,
  ParceleMaisEnvironment,
  BankAccountType,
  DisbursementModel,
  ParceleMaisApiError,
  type Establishment,
} from '@twila/parcelemais';

const client = new ParceleMaisClient({
  clientId: process.env.PARCELEMAIS_CLIENT_ID!,
  clientSecret: process.env.PARCELEMAIS_CLIENT_SECRET!,
  environment: ParceleMaisEnvironment.Staging,
});

export async function createEstablishment() {
  const { establishmentId } = await client.establishments.create({
    document: '12345678000199',
    legalName: 'Loja Centro LTDA',
    tradeName: 'Loja Centro',
    disbursementModel: DisbursementModel.EstablishmentChain,
    owner: {
      name: 'Maria Souza',
      email: 'maria@loja.com.br',
      phone: '+5511999998888',
    },
    bankAccount: {
      bankNumber: '341',
      agencyNumber: '1234',
      accountNumber: '56789',
      accountDigit: '0',
      accountType: BankAccountType.Current,
    },
    address: {
      street: 'Rua Exemplo',
      number: '100',
      district: 'Centro',
      city: 'São Paulo',
      state: 'SP',
      zipCode: '01310100',
    },
  });

  // Persist establishmentId — it's required to read, edit or change the status of the establishment later.
  return establishmentId;
}

export async function getEstablishment(establishmentId: string): Promise<Establishment | null> {
  try {
    return await client.establishments.get(establishmentId);
  } catch (error) {
    if (error instanceof ParceleMaisApiError && error.statusCode === 404) {
      console.warn(`Loja ${establishmentId} não pertence a este parceiro.`);
      return null;
    }
    throw error;
  }
}

export async function listActiveEstablishments(tradeName?: string) {
  return client.establishments.list({ tradeName, isActive: true });
}

export async function renameEstablishment(establishmentId: string) {
  await client.establishments.update(establishmentId, {
    tradeName: 'Loja Centro Matriz',
  });
}

export async function moveDisbursementToEstablishmentAccount(establishmentId: string) {
  // The establishment needs its own bank account before it can receive the disbursement.
  await client.establishments.updateBankAccount(establishmentId, {
    bankNumber: '341',
    agencyNumber: '1234',
    accountNumber: '56789',
    accountDigit: '0',
    accountType: BankAccountType.Current,
  });

  await client.establishments.update(establishmentId, {
    tradeName: 'Loja Centro Matriz',
    disbursementModel: DisbursementModel.Establishment,
  });
}

export async function closeEstablishment(establishmentId: string) {
  try {
    await client.establishments.deactivate(establishmentId);
  } catch (error) {
    if (error instanceof ParceleMaisApiError && error.statusCode === 404) {
      console.warn(`Loja ${establishmentId} não pertence a este parceiro.`);
      return;
    }
    throw error;
  }
}

export async function reopenEstablishment(establishmentId: string) {
  await client.establishments.activate(establishmentId);
}

File: examples/go/establishments.go
package main

import (
	"context"
	"errors"
	"fmt"
	"log"

	parcelemais "github.com/Twila-Digital/twila-parcelemais-go-sdk"
)

func createEstablishment(ctx context.Context, client *parcelemais.Client) string {
	created, err := client.Establishments.Create(ctx, parcelemais.CreateEstablishmentRequest{
		Document:          "12345678000199",
		LegalName:         "Loja Centro LTDA",
		TradeName:         "Loja Centro",
		DisbursementModel: parcelemais.DisbursementModelEstablishmentChain,
		Owner: parcelemais.EstablishmentOwner{
			Name:  "Maria Souza",
			Email: "maria@loja.com.br",
			Phone: "+5511999998888",
		},
		BankAccount: parcelemais.EstablishmentBankAccount{
			BankNumber:    "341",
			AgencyNumber:  "1234",
			AccountNumber: "56789",
			AccountDigit:  "0",
			AccountType:   parcelemais.BankAccountTypeCurrent,
		},
		Address: parcelemais.EstablishmentAddress{
			Street:   "Rua Exemplo",
			Number:   "100",
			District: "Centro",
			City:     "São Paulo",
			State:    "SP",
			ZipCode:  "01310100",
		},
	})
	if err != nil {
		log.Fatalf("create establishment: %v", err)
	}

	// Persist EstablishmentID — it's required to read, edit or change the status of the establishment later.
	fmt.Println("loja criada:", created.EstablishmentID)
	return created.EstablishmentID
}

func getEstablishment(ctx context.Context, client *parcelemais.Client, establishmentID string) *parcelemais.Establishment {
	establishment, err := client.Establishments.Get(ctx, establishmentID)
	if err == nil {
		return establishment
	}

	var apiErr *parcelemais.APIError
	if errors.As(err, &apiErr) && apiErr.StatusCode == 404 {
		log.Printf("loja %s não pertence a este parceiro", establishmentID)
		return nil
	}

	log.Fatalf("get establishment: %v", err)
	return nil
}

func listActiveEstablishments(ctx context.Context, client *parcelemais.Client, tradeName string) []parcelemais.Establishment {
	isActive := true

	establishments, err := client.Establishments.List(ctx, parcelemais.ListEstablishmentsRequest{
		TradeName: tradeName,
		IsActive:  &isActive,
	})
	if err != nil {
		log.Fatalf("list establishments: %v", err)
	}

	return establishments
}

func renameEstablishment(ctx context.Context, client *parcelemais.Client, establishmentID string) {
	if err := client.Establishments.Update(ctx, establishmentID, parcelemais.UpdateEstablishmentRequest{
		TradeName: "Loja Centro Matriz",
	}); err != nil {
		log.Fatalf("update establishment: %v", err)
	}
}

func moveDisbursementToEstablishmentAccount(ctx context.Context, client *parcelemais.Client, establishmentID string) {
	// The establishment needs its own bank account before it can receive the disbursement.
	if err := client.Establishments.UpdateBankAccount(ctx, establishmentID, parcelemais.EstablishmentBankAccount{
		BankNumber:    "341",
		AgencyNumber:  "1234",
		AccountNumber: "56789",
		AccountDigit:  "0",
		AccountType:   parcelemais.BankAccountTypeCurrent,
	}); err != nil {
		log.Fatalf("update establishment bank account: %v", err)
	}

	disbursementModel := parcelemais.DisbursementModelEstablishment

	if err := client.Establishments.Update(ctx, establishmentID, parcelemais.UpdateEstablishmentRequest{
		TradeName:         "Loja Centro Matriz",
		DisbursementModel: &disbursementModel,
	}); err != nil {
		log.Fatalf("update establishment: %v", err)
	}
}

func closeEstablishment(ctx context.Context, client *parcelemais.Client, establishmentID string) {
	err := client.Establishments.Deactivate(ctx, establishmentID)
	if err == nil {
		return
	}

	var apiErr *parcelemais.APIError
	if errors.As(err, &apiErr) && apiErr.StatusCode == 404 {
		log.Printf("loja %s não pertence a este parceiro", establishmentID)
		return
	}

	log.Fatalf("deactivate establishment: %v", err)
}

func reopenEstablishment(ctx context.Context, client *parcelemais.Client, establishmentID string) {
	if err := client.Establishments.Activate(ctx, establishmentID); err != nil {
		log.Fatalf("activate establishment: %v", err)
	}
}
